add mixed data type support for LayerNorm

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81852
* __->__ #81851
* #84405
* #84404

1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
add mixed data type support for LayerNorm

1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

ezyang commented:
test failure looks real

```
2022-07-22T03:41:59.6289214Z =================================== FAILURES ===================================
2022-07-22T03:41:59.6290167Z ______ TestCommonCPU.test_python_ref__refs_native_layer_norm_cpu_float32 _______
2022-07-22T03:41:59.6290866Z [gw1] linux -- Python 3.7.13 /opt/conda/bin/python
2022-07-22T03:41:59.6293201Z Traceback (most recent call last):
2022-07-22T03:41:59.6293527Z   File "/var/lib/jenkins/workspace/test/test_ops.py", line 496, in test_python_ref
2022-07-22T03:41:59.6293933Z     self._ref_test_helper(lambda: TorchRefsMode.push(strict=True), device, dtype, op)
2022-07-22T03:41:59.6294270Z   File "/var/lib/jenkins/workspace/test/test_ops.py", line 479, in _ref_test_helper
2022-07-22T03:41:59.6294568Z     self.assertTrue(ref_distance <= torch_distance, msg=msg)
2022-07-22T03:41:59.6294920Z   File "/opt/conda/lib/python3.7/unittest/case.py", line 705, in assertTrue
2022-07-22T03:41:59.6295169Z     raise self.failureException(msg)
2022-07-22T03:41:59.6295797Z AssertionError: tensor(False) is not true : Reference result was farther (1.7186287103232445e-05) from the precise computation than the torch result was (1.666915112918943e-05)!
2022-07-22T03:41:59.6536222Z 
```

ezyang commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -53,12 +50,12 @@ void LayerNormKernelImplInternal(
-      T_ACC mean_val;
-      T_ACC rstd_val;
+      T mean_val;
+      T rstd_val;
```
is the loss of ACC here intentional? IT isn't obvious to me that this template will never be instantiated with half precision type.

mingfeima replied:
this PR adds a explicit specification of `<BFloat16>` for the base template of `template <typename T>LayerNormKernelImplInternal` because of the need to dispatch to `layer_norm_kernel_mixed_type` when input is `BFloat16` and params is `BFloat16` or `float`.

So when input is `BFloat16`, it won't go to the base template of `LayerNormKernelImplInternal`. Previously `BFloat16` will also go to the base template so we need to use `T_ACC` in it. Now `T_ACC` would not been needed since `T` will only be float or double.

ezyang commented on aten/src/ATen/native/cpu/layer_norm_kernel.cpp:
```diff
@@ -96,10 +181,13 @@ void LayerNormKernelImpl(
+  DCHECK(!beta.defined() || beta.numel() == N);
   AT_DISPATCH_FLOATING_TYPES_AND(at::ScalarType::BFloat16, X.scalar_type(),
       "LayerNormKernelImpl", [&]() {
     LayerNormKernelImplInternal<scalar_t>(
```
and it indeed it looks like it's being run at low precision here

mingfeima replied:
Now the dispatch pattern is like:

* input is float, param is float: it will call `template <typename T> LayerNormKernelImplInternal` and `T` is float.

* input is bfloat16, param is bfloat16, it will call `template <>LayerNormKernelImplInternal<BFloat16>`, and it will go to `template <typename param_t>layer_norm_kernel_mixed_type` and `param_t` is BFloat16.

* input is bfloat16, param is float, it will call `template <>LayerNormKernelImplInternal<BFloat16>`, and it will go to `template <typename param_t>layer_norm_kernel_mixed_type` and `param_t` is float.

Sorry for the late response, working on some difficult tasks last few days.

malfet reviewed (changes requested):
As Ed commented 10 days ago, test_python_ref__refs_native_layer_norm_cpu_float32 seems real and related to the PR in question, isn't it?

mingfeima commented:
> As Ed commented 10 days ago, test_python_ref__refs_native_layer_norm_cpu_float32 seems real and related to the PR in question, isn't it?

Yes, this failure is solid. This PR has dependency on this performance regression on https://github.com/pytorch/benchmark/issues/1099. Will have the CI fixed when handling this performance regression.

Sorry for the late response, I have to deal with some difficult optimizations on pyg last few days. Will fix the issues on this stack asap.

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima commented:
Fix CI failure on `test_python_ref__refs_native_layer_norm_cpu_float32 `.

ezyang commented:
@pytorchbot merge

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "distributed failure looks unrelated, test was previously disabled"

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for LayerNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16 such as:
```
  x = torch.randn(n, t, c).bfloat16()
  ln = nn.LayerNorm(c).bfloat16()
  y = ln(x)
```
The input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@pytorchbot merge
